File: lab10/src/by/bntu/fitr/minigame/domain/Creature.java
package by.bntu.fitr.minigame.domain;

abstract public class Creature implements Buff {
    private int health;
    private int strength;
    private int agility;
    private int intelligence;

    public Creature(int health, int strength, int agility, int intelligence) {
        this.health = health;
        this.strength = strength;
        this.agility = agility;
        this.intelligence = intelligence;
    }

    abstract protected int getWeaponDamage();

    public int dealDamage() {
        return buff() + getWeaponDamage();
    }

    public int getHealth() {
        return health;
    }

    public int getStrength() {
        return strength;
    }

    public int getAgility() {
        return agility;
    }

    public int getIntelligence() {
        return intelligence;
    }

    @Override
    public String toString() {
        return "health=" + health +
                ", strength=" + strength +
                ", agility=" + agility +
                ", intelligence=" + intelligence + ",";
    }
}

File: lab10kt/src/main/kotlin/by/bntu/fitr/povt/korotkevich/model/Creature.kt
package by.bntu.fitr.povt.korotkevich.model

import by.bntu.fitr.povt.korotkevich.model.Role.*

abstract class Creature {

    abstract val id: String
    abstract val health: Int
    abstract val strength: Int
    abstract val agility: Int
    abstract val intelligence: Int

    abstract fun getSelfDamage(): Int

    override fun toString() = StringBuilder().apply {
        append("health: $health, strength: $strength, ")
        append("agility: $agility, intelligence: $intelligence")
    }.toString()
}

fun Creature.buff(role: Role)= (when(role) {
    KNIGHT -> health
    SORCERER -> intelligence
    RANGER -> agility
    TANK -> strength
    THIEF -> agility
} * role.buffPercent).toInt()
